[#27] 게시글 좋아요 기능 구현

JIWON27 opened the pull request:
### 설명
게시글 좋아요 구현

### TODO 
- [x] 게시글 좋아요/ 좋아요 취소 기능
- [x] 게시글 좋아요 개수 조회
- [x] 좋아요 누른 게시글 최신순 조회 기능
- [x] 게시글 삭제 시 redis에서도 해당 게시글 기록 삭제

JIWON27 committed:
feat: PostLike 엔티티, repository 추가

JIWON27 committed:
feat: 특정 게시글에 좋아요가 존재하는지 확인하는 예외 추가

JIWON27 committed:
feat: PostLikeService 구현

JIWON27 committed:
feat: PostLikeController 구현

JIWON27 committed:
feat: PostIds In절 조회 기능 추가

JIWON27 committed:
feat: 게시글 상세 조회 시 좋아요 개수 필드 추가

JIWON27 committed:
feat: 게시글 삭제 시 좋아요 관련 기록 삭제하도록 추가

JIWON27 committed:
feat: PostLike HTTP 요청 파일

JIWON27 committed:
feat: Redis 트랜잭션 기능 추가

JIWON27 committed:
feat: PostLikeService에 @Transactional 추가

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/service/PostLikeService.java:
```diff
@@ -0,0 +1,132 @@
+public class PostLikeService {
+    private final String USER_REDIS_PREFIX = "user";
+    private final String POST_REDIS_PREFIX = "post";
+    private final RedisTemplate<String, Long> restTemplate;
```
restTemplate 오타가 있는 것 같네요~

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/service/PostLikeService.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    private void validateExistPost(Long postId){
+        postRepository.findById(postId).orElseThrow(() -> new NotFoundPostException());
```
여기서 exists 함수를 사용하는 건 어떨까요?

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/service/PostLikeService.java:
```diff
@@ -0,0 +1,132 @@
+
+        restTemplate.opsForZSet().add(userKey, postId, score);
+        restTemplate.opsForSet().add(postKey, user.getId());
+        return PostResponse.Like.from(count(postId), true);
```
return 문을 맨 아래 하나만 두고, 좋아요 한 기록이 없을 경우를 if문 안에 넣어 수행하도록 작성하시는 건 어떨까요?? 리턴 문이 한 눈에 보여서 알기 쉬워질 것 같아요!

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/service/PostLikeService.java:
```diff
@@ -0,0 +1,132 @@
+        }
+        List<Post> posts = postRepository.findAllByIdIn(postIds);
+
+        List<Post> filterPosts = posts.stream()
+                                      .filter(post -> post.getUser().getId() != user.getId())
+                                      .collect(Collectors.toList());
```
일관성을 위해, 내가 작성한 게시글에 좋아요를 눌렀더라도 함께 반환해주는 것이 좋을 것 같아요

rnignon commented on module-api/src/main/java/com/trybe/moduleapi/post/exception/AlreadyLikePostException.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.http.HttpStatus;
+
+public class AlreadyLikePostException extends BusinessException {
+    public AlreadyLikePostException() {
```
Duplicated~~Exception으로 통일해요!!

rnignon reviewed (changes requested):
수고하셨습니다~ 👩‍🔬

JIWON27 committed:
refactor: 좋아요 개수 likeCount로 변경

JIWON27 committed:
refactor: 불필요한 주석 삭제 및 delete(..) 내부 삭제 메서드 호출 순서 변경

JIWON27 committed:
refactor: ProductLikeService 코드 수정
- restTemplate -> redisTemaplate로 수정
- validateExistPost(..) existsById 메서드 사용하도록 수정
- addLike/removeLike 메서드 좋아요 기록에 따른 if문 수정

JIWON27 committed:
refactor: 좋아요 추가 컨트롤러 메서드명 addLike로 수정
